### DISCUSSION TITLE
Nextjs example blog subdirectory pages / subpages

### DISCUSSION THREAD

**Discussion by Prottoy2938** [vercel/next.js#19307] [Ideas]
I'm following this [nextjs blog example](https://github.com/vercel/next.js/tree/canary/examples/blog-starter-typescript).

To create a new page, I have to create a new **.md** file in `/_posts` directory and it will create a new page with that file name.
Everything works well in there.

But is it possible to create a subdirectory/ subpages, as we do in the default **nextjs** layout. Where we create a new folder in the `/pages` directory and nextjs serves pages under that subdirectory.

So is it possible to configure something like that in that markdown blog example?

@timneutkens

**Comment by xypnox** [comment]
@Prottoy2938 Did you figure this out?

**Reply by Prottoy2938** [reply]
No

**Reply by xypnox** [reply]
Quoting @davecaruso 's solution from discord:


@xypnox i wrote a recursive file reader that returns a flat array of files, then i map the paths to use that catch-all thing you sent earlier. this works because the filename is [[...slug]].tsx

![image](https://user-images.githubusercontent.com/25076171/102654645-e6eb1d00-4196-11eb-9b60-0b4ae42e7184.png)

then i just have this as the DOCS_ROOT folder (the existing pages are me documenting how im building my website)

![image](https://user-images.githubusercontent.com/25076171/102654659-efdbee80-4196-11eb-91e8-9fca48c2adbb.png)

then i have getStaticProps and the actual page component, but you should be able to get that setup already, if you want my implementation just ask.
and it does.

![image](https://user-images.githubusercontent.com/25076171/102654684-fe2a0a80-4196-11eb-9fba-9ff98fd8cb37.png)

actually i just realized the index.mdx would not load using this, but rather it would be at the folder/index url

@xypnox alright i fixed all that but its a bit more complex so heres the entire thing https://hastebin.com/jugolazaje.typescript

**Reply by xypnox** [reply]
The file's contents:

```typescript
import fs from 'fs-extra';
import hydrate from 'next-mdx-remote/hydrate';
import renderToString from 'next-mdx-remote/render-to-string';
import dynamic from 'next/dynamic';
import Head from 'next/head';
import Link from 'next/link';
import path from 'path';
import { GetStaticPaths, InferGetStaticPropsType } from 'next';
import matter from 'gray-matter'
import rehypeHighlight from 'util/rehype-shiki';
import { ParsedUrlQuery } from 'querystring';
import 'components/docs/DocPage.scss';

// File path for docs root
const DOCS_ROOT = path.join(process.cwd(), '/docs');

// Custom Components for MDX
const components = {
  a: Link,
  head: Head,

  ArtifactDemoDate: dynamic(() => import('components/docs/ArtifactDemo').then(m => m.ArtifactDemoDate as any)),
  ArtifactDemoID: dynamic(() => import('components/docs/ArtifactDemo').then(m => m.ArtifactDemoID as any)),

  StylingSampleCompontent: dynamic(() => import('components/docs/Component'))
} as any;

// Actual React Component
type PostPageProps = InferGetStaticPropsType<typeof getStaticProps>

export default function DocPage({ source, meta }: PostPageProps) {
  const content = hydrate(source, { components })

  return <div className="root">
    <Head>
      <link href="https://fonts.googleapis.com/css2?family=Source+Code+Pro:ital@0;1&display=swap" rel="stylesheet" />
    </Head>
    <div className="paper">
      {content}
    </div>
  </div>
}

// Define a cache that will map the slug to the actual path. This is used to work around
// the index.mdx issue that I'm having
const pageFileCache: Record<string, string> = {};

// Get Static Props, handles markdown processing.
const languageNameMap: Record<string, string> = {
  gitignore: 'ini'
}

export const getStaticProps = async ({ params }: any) => {
  // Retrieve full path from a cache. Generate cache if it doesnt exist.
  const slugPath = params.slug.join('/');
  if (!pageFileCache[slugPath]) {
    await getStaticPaths({});
  }
  const markdownFile = pageFileCache[slugPath];
  const source = await fs.readFile(markdownFile)

  // Use frontmatter
  const { content, data } = matter(source);

  // Do an extra transformation that lets me include code files by filename
  const processedContent = content
    .replace(
      /^\s*```include\s+(.*)\s*```\s*$/gm,
      (fullText: string, file) => {
        const ext = file.split('.').pop();
        const langname = languageNameMap[ext] || ext;
        return '``````' + langname + '\n' + fs.readFileSync(path.join(process.cwd(), file)) + '\n``````\n'
      }
    )

  // Pass through MDX
  const mdxSource = await renderToString(processedContent, {
    components,
    mdxOptions: {
      rehypePlugins: [
        [rehypeHighlight, { theme: 'util/one-dark-pro.json' }]
      ],
    },
    scope: data,
  });

  // Return Props
  return {
    props: {
      source: mdxSource,
      meta: data,
    },
  }
}

// Recursivly scan the documentation folder. This includes any extra resolution magic
// such as index.mdx acting as the folder index
async function getPathList(folder: string): Promise<{ params: ParsedUrlQuery }[]> {
  return (await Promise.all((await fs.readdir(folder)).map(async(file) => {
    const joined = path.join(folder, file);
    if ((await fs.stat(joined)).isDirectory()) {
      return getPathList(joined);
    }
    if (joined.endsWith('.mdx')) {
      const alteredPath = joined
        .slice(DOCS_ROOT.length + 1)
        .replace(/\.mdx$/, '')
        .replace(/\/index$/, '');
      
      pageFileCache[alteredPath] = joined;

      return {
        params: {
          slug: alteredPath.split('/'),
        },
      };
    } else {
      return null;
    }
  }))).flat().filter(x => !!x) as { params: ParsedUrlQuery }[];
}

// Get static paths being a wrapper around the getPathList
export const getStaticPaths: GetStaticPaths = async () => {
  const paths = await getPathList(DOCS_ROOT);

  return {
    paths,
    fallback: false,
  }
}

```

**Comment by xypnox** [comment]
@Prottoy2938  

I reached a somewhat stable code for nested files including component to display the nested structure. 

I have deployed a sample version here: https://next-nested.vercel.app/ and the code is at: https://github.com/xypnox/next-nested

Hope this helps!

**Reply by Prottoy2938** [reply]
Thanks, mate. I think you should create a pull request in the [nextjs examples](https://github.com/vercel/next.js/tree/canary/examples) section on github and add this new example. 

You could title this as `blog-starter-subpages`

This would be quite helpful to others too.

**Reply by mattfelten** [reply]
After a couple tries I got this working too. This has been the only example I've found that works. Gonna try to add MDX support to it right now
